pkg/client/unversioned/remotecommand/v3.go

```diff
@@ -0,0 +1,121 @@
+/*
+Copyright 2016 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package remotecommand
+
+import (
+	"encoding/json"
+	"fmt"
+	"net/http"
+	"sync"
+
+	"k8s.io/kubernetes/pkg/api"
+	"k8s.io/kubernetes/pkg/util/httpstream"
+	"k8s.io/kubernetes/pkg/util/runtime"
+)
+
+// streamProtocolV3 implements version 3 of the streaming protocol for attach
+// and exec. This version adds support for resizing the container's terminal.
+type streamProtocolV3 struct {
+	*streamProtocolV2
+
+	resizeStream httpstream.Stream
+}
+
+var _ streamProtocolHandler = &streamProtocolV3{}
+
+func newStreamProtocolV3(options StreamOptions) streamProtocolHandler {
+	return &streamProtocolV3{
+		streamProtocolV2: newStreamProtocolV2(options).(*streamProtocolV2),
+	}
+}
+
+func (p *streamProtocolV3) createStreams(conn streamCreator) error {
+	// set up the streams from v2
+	if err := p.streamProtocolV2.createStreams(conn); err != nil {
+		return err
+	}
+
+	// set up resize stream
+	if p.Tty {
+		headers := http.Header{}
+		headers.Set(api.StreamType, api.StreamTypeResize)
+		var err error
+		p.resizeStream, err = conn.CreateStream(headers)
+		if err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func (p *streamProtocolV3) handleResizes() {
+	if p.resizeStream == nil {
+		return
+	}
+
+	go func() {
+		defer runtime.HandleCrash()
+
+		encoder := json.NewEncoder(p.resizeStream)
+		for {
+			size := p.TerminalSizeQueue.Next()
+			if size == nil {
+				return
+			}
+			if err := encoder.Encode(&size); err != nil {
+				runtime.HandleError(err)
+			}
+		}
+	}()
+}
+
+func (p *streamProtocolV3) stream(conn streamCreator) error {
+	// set up all the streams first
+	if err := p.createStreams(conn); err != nil {
+		return err
+	}
+
+	// now that all the streams have been created, proceed with reading & copying
+
+	// always read from errorStream
+	errorChan := p.setupErrorStreamReading()
+
+	p.handleResizes()
+
+	if p.Stdin != nil && p.Tty {
+		// Some images, for some reason, don't refresh the display when you attach and resize.
```

Comment:
Yes it affects Docker too. It's only certain images (same behavior for both)

On Tuesday, July 5, 2016, Clayton Coleman <notifications@github.com> wrote:

> In pkg/client/unversioned/remotecommand/v3.go
> <https://github.com/kubernetes/kubernetes/pull/25273#discussion_r69647111>
> :
>
> > +
> > +func (p *streamProtocolV3) stream(conn streamCreator) error {
> > +	// set up all the streams first
> > +	if err := p.createStreams(conn); err != nil {
> > +		return err
> > +	}
> > +
> > +	// now that all the streams have been created, proceed with reading & copying
> > +
> > +	// always read from errorStream
> > +	errorChan := p.setupErrorStreamReading()
> > +
> > +	p.handleResizes()
> > +
> > +	if p.Stdin != nil && p.Tty {
> > +		// Some images, for some reason, don't refresh the display when you attach and resize.
>
> I feel like this needs to be optional and we need to dig into why this
> happens. Does this affect Docker?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/25273/files/d7dfeb6149695cab3c01f649fcf43d8d9205d98e#r69647111>,
> or mute the thread
> <https://github.com/notifications/unsubscribe/AAABYkLnWD2DKqGV3WMJKUyb_1vIoYVsks5qSta3gaJpZM4IZGqq>
> .
>